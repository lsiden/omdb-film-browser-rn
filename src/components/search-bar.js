// Debounces react-native-elements.SearchBar

import React from "react"
import { connect } from "react-redux"
import PropTypes from "prop-types"
import { SearchBar } from "react-native-elements"
import { Platform } from "react-native"

import { fetchQuery } from "actions/fetch"

export class searchBar extends React.Component {
  static propTypes = {
    dispatchQuery: PropTypes.func.isRequired
  }

  constructor(props) {
    super(props)
    this.state = { query: "" }
    this.onChangeText = this.onChangeText.bind(this)
    this.onTimeout = this.onTimeout.bind(this)
    this.onCancel = this.onCancel.bind(this)
  }

  onTimeout() {
    const { query } = this.state
    this.timeoutId = null

    if (query) {
      this.props.dispatchQuery(query)
    }
  }

  onChangeText(query) {
    this.setState({ query })

    if (this.timeoutId) {
      clearTimeout(this.timeoutId)
    }
    this.timeoutId = setTimeout(this.onTimeout, this.props.delay)
  }

  onCancel() {
    this.setState({ query: "" })
    this.ref.clear()
  }

  render() {
    return (
      <SearchBar
        placeholder="Title..."
        onChangeText={this.onChangeText}
        onCancel={this.onCancel}
        clearIcon={{ color: "#86939e", name: "close" }}
        platform={Platform.OS}
      />
    )
  }
}

searchBar.propTypes = {
  dispatchQuery: PropTypes.func.isRequired,
  delay: PropTypes.number
}

searchBar.defaultProps = {
  delay: 300
}

export default connect(null, dispatch => ({
  dispatchQuery: query => {
    dispatch(fetchQuery(query))
  }
}))(searchBar)
