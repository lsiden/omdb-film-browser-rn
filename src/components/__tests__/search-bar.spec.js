import React from "react"
import renderer from "react-native-test-utils"

import { searchBar as SearchBar } from "../search-bar"

let dispatchQuery

jest.useFakeTimers()

const defaultProps = () => {
  dispatchQuery = jest.fn()
  return { dispatchQuery }
}
const createWrapper = (props = {}) =>
  renderer(<SearchBar {...{ ...defaultProps(), ...props }} />)

it("renders", () => {
  const wrapper = createWrapper()
  expect(wrapper.toJSON()).toMatchSnapshot()
})

it("responds to change in input only after delay", () => {
  const delay = 555
  const wrapper = createWrapper({ delay })
  wrapper.instance().onChangeText("foo")
  expect(setTimeout).toHaveBeenCalledTimes(1)
  expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), delay)
})

test("onTimeout() invokes dispatchQuery()", () => {
  const wrapper = createWrapper()
  const instance = wrapper.instance()
  instance.setState({ query: "a query" })
  instance.onTimeout()
  expect(dispatchQuery).toHaveBeenCalled()
})

test("onTimeout() does not invoke dispatchQuery() if state.query is empty", () => {
  const wrapper = createWrapper()
  wrapper.instance().onTimeout()
  expect(dispatchQuery).not.toHaveBeenCalled()
})
