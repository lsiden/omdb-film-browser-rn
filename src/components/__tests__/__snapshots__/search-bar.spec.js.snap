// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#393e42",
        "borderBottomColor": "#000",
        "borderBottomWidth": 1,
        "borderTopColor": "#000",
        "borderTopWidth": 1,
      },
      false,
      undefined,
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    onCancel={[Function]}
    onChangeText={[Function]}
    placeholder="Title..."
    placeholderTextColor="#86939e"
    style={
      Array [
        Object {
          "backgroundColor": "#303337",
          "borderRadius": 3,
          "color": "#86939e",
          "fontSize": 17.5,
          "height": 30,
          "margin": 8,
          "overflow": "hidden",
          "paddingLeft": 26,
          "paddingRight": 19,
        },
        false,
        false,
        false,
        undefined,
        false,
        Object {
          "paddingRight": 30,
        },
      ]
    }
    underlineColorAndroid="transparent"
  />
  <Text
    accessible={true}
    allowFontScaling={false}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#86939e",
          "fontSize": 16,
        },
        Array [
          Object {
            "backgroundColor": "transparent",
            "position": "absolute",
            "top": 15.5,
          },
          Object {
            "left": 16,
          },
          undefined,
        ],
        Object {
          "fontFamily": "Material Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      ]
    }
  >
    
  </Text>
  <Text
    accessible={true}
    allowFontScaling={false}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Array [
        Object {
          "color": "#86939e",
          "fontSize": 16,
        },
        Array [
          Object {
            "backgroundColor": "transparent",
            "position": "absolute",
            "top": 15.5,
          },
          Object {
            "right": 16,
          },
          undefined,
        ],
        Object {
          "fontFamily": "Material Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      ]
    }
  >
    
  </Text>
</View>
`;
